Sources/Datadog/Core/SessionTracer.swift

```diff
@@ -0,0 +1,54 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import _Datadog_Private
+
+// TODO: RUMM-300 This should probably be a part of DDTracer
+public extension Datadog {
+    internal typealias RequestInterceptor = (URLRequest) -> URLRequest
+    internal typealias TaskObserver = (URLSessionTask) -> Void
+
+    static func trace(_ session: URLSession) throws {
+        let requestInterceptor: RequestInterceptor = { originalRequest in
+            // TODO: RUMM-300 Set tracer HTTP header values
+            return originalRequest
+        }
+        let taskObserver: TaskObserver = { task in
+            let observation: NSKeyValueObservation
+            var oldState: URLSessionTask.State? = nil
+            // URLSessionTask.state does NOT support NSKeyValueObservedChange
+            observation = task.observe(\.state) { observedTask, _ in
+                let newState = observedTask.state
+                switch (oldState, newState) {
```

Comment:
Let's decouple this transition logic to separate function, so we can unit test all possible transitions.

Reply:
🎗️ I left a `todo` for now, we can implement it when we plug `tracer`